### PR TITLE
vmware: fix vm snapshot with datastore cluster, drs

### LINKED ISSUE #6595: VMware VM snapshot failure in case of storage pool type datastore cluster
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
VMware, VMsnapshot, Datastore Cluster
~~~

##### CLOUDSTACK VERSION

~~~
4.15.0 and above
~~~

##### SUMMARY

VM snapshot is failing on some VMs for which disks are created in a storage pool of type “Datastore Cluster“ in VMware.
The issue is happening in a specific case as mentioned below in the reproduction steps.

##### STEPS TO REPRODUCE

~~~

1. Create a CloudStack environment with a storage pool of type “Datastore Cluster“ (Assume Datastore cluster has two child pools in it “pool A“ and “pool B“) and make sure DRS is enabled on the datastore cluster on vCenter.
2. Create a VM with a root disk and a data disk.
3. Create a scenario where both root and data disks are created in the same child storage pool in datastore cluster, say “pool A“
4. Let the DRS trigger on vCenter and moved the data disk from “pool A“ to “pool B“
5. Now perform VM snapshot operation from CloudStack which will fail with the below error.

```
2022-08-01 06:14:46,623 DEBUG [c.c.v.s.VMSnapshotManagerImpl] (Work-Job-Executor-6:ctx-844c99dc job-73/job-74 ctx-5482d2e8) (logid:df1e2c15) Failed to create vm snapshot: 6
com.cloud.utils.exception.CloudRuntimeException: Creating VM snapshot: i-2-5-VM_VS_20220801061435 failed due to File [783bcfa34486393f8b597e3af109a56b] i-2-5-VM was not found

``` 

Summary: The issue is happening when the CloudStack’s database has the details for two volumes pool_id and chain_info pointing to the same child storage pool but on vCenter, they are in different child storage pools. In other cases, even though DRS moved the volumes around, CloudStack can still take VM snapshot.

Workaround: Stop and start the VM. CloudStack will reconcile the volumes information during that process.
~~~

##### EXPECTED RESULTS

~~~
VM snapshot should be successful and CloudStack should update the new volumes information
~~~

##### ACTUAL RESULTS

~~~
VM snapshot operation is failing in the specific case as mentioned above in reproduction steps
~~~

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #6595 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test matrix

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test centos7 vmware-65u2

File: engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/vmsnapshot/DefaultVMSnapshotStrategy.java [hunk 1]
```diff
@@ -325,12 +329,24 @@ protected void finalizeRevert(VMSnapshotVO vmSnapshot, List<VolumeObjectTO> volu
...
+            if (StringUtils.isNotEmpty(volume.getPath())) {
                 volumeVO.setPath(volume.getPath());
-                volumeVO.setVmSnapshotChainSize(volume.getSize());
-                volumeDao.persist(volumeVO);
             }
+            if (StringUtils.isNotEmpty(volume.getChainInfo())) {
+                volumeVO.setChainInfo(volume.getChainInfo());
+            }
+            volumeVO.setVmSnapshotChainSize(volume.getSize());
+            volumeDao.persist(volumeVO);
```
Comment (DaanHoogland) [inline engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/vmsnapshot/DefaultVMSnapshotStrategy.java:349]: More of a general remark than about this piece of code; can these lines be extracted and tested in unit tests. (extracted is maybe not needed in this case, but unit tests would be nice)

**Comment by DaanHoogland** [comment]
is this ready for review @shwstppr ?

**Comment by shwstppr** [comment]
@DaanHoogland yes

**Review by DaanHoogland — APPROVED** [review]
clgtm, several bits could be extracted to extra methods and unit tests added. I'm not sure to what extend this would be useful, given the code coverage reports in between. It seems most of those reports refer to old moved code, which doesn´t mean we could add coverage, but is somewhat a false positive.

File: engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/vmsnapshot/DefaultVMSnapshotStrategy.java (hunk 1 above)
Comment (shwstppr) [reply engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/vmsnapshot/DefaultVMSnapshotStrategy.java:349]: @DaanHoogland added unit test

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by rohityadavcloud — APPROVED** [review]

**Comment by rohityadavcloud** [comment]
@blueorangutan test centos7 vmware-70u2

**Comment by rohityadavcloud** [comment]
@shwstppr is this good to merge or are we looking for another review or manual test?

**Comment by shwstppr** [comment]
@rohityadavcloud this is being manually tested by @NuxRo

**Review by NuxRo — APPROVED** [review]
LGTM

**Pull request merged**